src/main/java/org/truffleruby/interop/OutgoingForeignCallNode.java

```diff
@@ -99,8 +103,98 @@ protected int objectId(Object receiver, String name, Object[] args,
             @Cached(value = "name", allowUncached = true) @Shared("name") String cachedName,
+            @CachedLibrary("receiver") InteropLibrary interop,
+            @Cached TranslateInteropExceptionNode translateInteropException,
+            @CachedContext(RubyLanguage.class) RubyContext context,
             @Cached InteropNodes.ReadArrayElementNode readNode) {
-        return readNode.execute(receiver, args[0]);
+        return at(receiver, name, args, cachedName, interop, translateInteropException, context, readNode);
+    }
+
+    @Specialization(
+            guards = {
+                    "name == cachedName",
+                    "cachedName.equals(AT)",
+                    "args.length == 1",
+                    "isBasicInteger(first(args))" },
+            limit = "1")
+    protected Object at(Object receiver, String name, Object[] args,
+            @Cached(value = "name", allowUncached = true) @Shared("name") String cachedName,
+            @CachedLibrary("receiver") InteropLibrary interop,
+            @Cached TranslateInteropExceptionNode translateInteropException,
+            @CachedContext(RubyLanguage.class) RubyContext context,
+            @Cached InteropNodes.ReadArrayElementNode readNode) {
+        try {
+            long size = interop.getArraySize(receiver);
+            long bound = (size != 0) ? size : 0;
+            if (((int) args[0]) < -bound || (int) args[0] >= bound) {
+                return nil;
+            }
+            return ((int) args[0] < 0)
+                    ? readNode.execute(receiver, size + (int) args[0])
+                    : readNode.execute(receiver, args[0]);
+        } catch (UnsupportedMessageException e) {
+            throw translateInteropException.execute(e);
+        }
+    }
+
+    @Specialization(
+            guards = {
+                    "name == cachedName",
+                    "cachedName.equals(FETCH)",
+                    "args.length == 1",
+                    "isBasicInteger(first(args))" },
+            limit = "1")
+    protected Object fetch(Object receiver, String name, Object[] args,
+            @Cached(value = "name", allowUncached = true) @Shared("name") String cachedName,
+            @CachedLibrary("receiver") InteropLibrary interop,
+            @Cached TranslateInteropExceptionNode translateInteropException,
+            @CachedContext(RubyLanguage.class) RubyContext context,
+            @Cached InteropNodes.ReadArrayElementNode fetchNode) {
+        try {
+            long size = interop.getArraySize(receiver);
+            long bound = (size != 0) ? size : 0;
+            if (((int) args[0]) < -bound || (int) args[0] >= bound) {
+                throw new RaiseException(
+                        context,
+                        context.getCoreExceptions().indexError("Index " + (int) args[0] +
+                                " outside of array bounds: -" + bound + "..." + bound, this));
```

Comment:
Actually, `InteropNodes.ReadArrayElementNode` already raises an IndexError if the index is out of bounds, so I think we don't need different logic for `fetch` here.

Reply:
But the exception message send by InteropNodes.ReadArrayElementNode is not the same as send by MRI in this case.  Maybe we need to change the ReadArraElementNode exception message.

Reply:
I think the exception message doesn't matter too much, I think having the same exception class is enough for most cases.

Reply:
Just realized that the message has a bug, the same as in JRuby:  if the array has 3 elements and we do
fetch(-5) the exception message is: "invalid array index -2".